File: apps/api/src/xtrusio_api/schemas/me.py
"""Response schema for GET /api/me."""

from __future__ import annotations

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, EmailStr

from ..models.platform_user import PlatformRole
from ..models.tenant_membership import TenantRole


class PlatformContext(BaseModel):
    role: PlatformRole
    is_active: bool


class TenantContext(BaseModel):
    id: UUID
    slug: str
    name: str
    role: TenantRole
    # Resolver-derived effective workspace permission keys for this tenant.
    # Additive: `role` is kept until the frontend migrates (P6b).
    permissions: list[str]


class PendingInvite(BaseModel):
    kind: Literal["platform", "tenant"]
    id: UUID
    tenant_id: UUID | None
    role: str  # widened union of platform_role and tenant_role; validated server-side


class MeResponse(BaseModel):
    user_id: UUID
    email: EmailStr
    platform: PlatformContext | None
    # Resolver-derived effective platform permission keys (empty if none).
    # Additive: `platform.role` is kept until the frontend migrates (P6b).
    platform_permissions: list[str]
    tenants: list[TenantContext]
    pending_invite: PendingInvite | None

File: packages/api-types/src/me.ts
// packages/api-types/src/me.ts
// Mirror of apps/api/src/xtrusio_api/schemas/me.py:MeResponse. The enum
// fields (`platform.role`, `tenants[].role`) are kept ADDITIVELY alongside
// the resolver-derived permission arrays so frontend consumers can migrate
// component-by-component. The enum fields will be removed in a later phase
// once every consumer reads permissions instead.

export type PlatformRole = "super_admin" | "admin" | "editor";
export type TenantRole = "owner" | "admin" | "editor" | "read_only";

/** Permission key as defined in apps/api/src/xtrusio_api/rbac/catalog.py. */
export type PermissionKey = string;

export type PlatformContext = {
  role: PlatformRole;
  is_active: boolean;
};

export type TenantContext = {
  id: string;
  slug: string;
  name: string;
  role: TenantRole;
  /** Resolver-derived effective workspace permission keys. */
  permissions: PermissionKey[];
};

export type PendingInvite = {
  kind: "platform" | "tenant";
  id: string;
  tenant_id: string | null;
  role: string;
};

export type MeResponse = {
  user_id: string;
  email: string;
  platform: PlatformContext | null;
  /** Resolver-derived effective platform permission keys (empty if none). */
  platform_permissions: PermissionKey[];
  tenants: TenantContext[];
  pending_invite: PendingInvite | null;
};
